Visual Studio(msvc) is warning

普段はmsys(gcc)で作業をしていて、Visual Studioでは
コンパイルが通るか程度しか確認を行っていませんでしたが
まずそうな警告が出ていました。

該当部分
```
1>player.cpp(78): warning C4930: 'std::lock_guard<std::mutex> lock(std::mutex)': プロトタイプされている関数が呼び出されませんでした (変数の定義が意図されていますか?)
```

全文
```
1>------ ビルド開始: プロジェクト: official, 構成: Release x64 ------
2>------ ビルド開始: プロジェクト: player_greedy, 構成: Release x64 ------
3>------ ビルド開始: プロジェクト: player_pseudo_alphabeta, 構成: Release x64 ------
2>greedy.cpp
2>raceState.cpp
1>course.cpp
3>pseudo_alphabeta.cpp
3>raceState.cpp
2>コード生成しています。
2>2 of 598 functions ( 0.3%) were compiled, the rest were copied from previous compilation.
3>コード生成しています。
2>  0 functions were new in current compilation
2>  9 functions had inline decision re-evaluated but remain unchanged
2>コード生成が終了しました。
3>0 of 962 functions ( 0.0%) were compiled, the rest were copied from previous compilation.
2>player_greedy.vcxproj -> C:\...\Software2017-18\x64\Release\player_greedy.exe
3>  0 functions were new in current compilation
3>  9 functions had inline decision re-evaluated but remain unchanged
3>コード生成が終了しました。
1>main.cpp
3>player_pseudo_alphabeta.vcxproj -> C:\...\Software2017-18\x64\Release\player_pseudo_alphabeta.exe
1>player.cpp
1>player.cpp(78): warning C4930: 'std::lock_guard<std::mutex> lock(std::mutex)': プロトタイプされている関数が呼び出されませんでした (変数の定義が意図されていますか?)
1>player.cpp(141): warning C4244: '-=': '__int64' から 'int' への変換です。データが失われる可能性があります。
1>player.cpp(263): warning C4244: '-=': '__int64' から 'int' への変換です。データが失われる可能性があります。
1>raceState.cpp
1>コード生成しています。
1>0 of 4291 functions ( 0.0%) were compiled, the rest were copied from previous compilation.
1>  0 functions were new in current compilation
1>  82 functions had inline decision re-evaluated but remain unchanged
1>コード生成が終了しました。
1>official.vcxproj -> C:\...\Software2017-18\x64\Release\official.exe
1>プロジェクト "official.vcxproj" のビルドが終了しました。
========== ビルド: 3 正常終了、0 失敗、0 更新不要、0 スキップ ==========
```

（ `using namespace std` をしているので当たり前ですが…）

修整します。

--- a/official/official.vcxproj
+++ b/official/official.vcxproj
@@ -72,7 +72,7 @@
   <PropertyGroup />
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>Disabled</Optimization>
       <SDLCheck>false</SDLCheck>
       <AdditionalOptions>/source-charset:utf-8 %(AdditionalOptions)</AdditionalOptions>
@@ -98,7 +98,7 @@
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>MaxSpeed</Optimization>
       <FunctionLevelLinking>true</FunctionLevelLinking>
       <IntrinsicFunctions>true</IntrinsicFunctions>

--- a/official/player.cpp
+++ b/official/player.cpp
@@ -6,6 +6,7 @@
 #include <thread>
 #include <chrono>
 #include <mutex>
+#include <cinttypes>
 #include <boost/optional.hpp>
 
 #include "course.hpp"
@@ -52,11 +53,12 @@ static void handShake(std::unique_ptr<boost::process::ipstream> in, std::promise
   p.set_value(make_pair(std::move(in), ans));
 }
 
-void sendToAI(std::unique_ptr<boost::process::opstream>&  toAI, std::shared_ptr<std::ofstream> stdinLogStream, const char *fmt, int value) {
-  int n = std::snprintf(nullptr, 0, fmt, value);
+template <class... Args>
+void sendToAI(std::unique_ptr<boost::process::opstream>&  toAI, std::shared_ptr<std::ofstream> stdinLogStream, const char *fmt, Args... args) {
+  int n = std::snprintf(nullptr, 0, fmt, args...);
   std::unique_ptr<char[]> cstr(new char[n + 2]);
   std::memset(cstr.get(), 0, n + 2);
-  std::snprintf(cstr.get(), n + 1, fmt, value);
+  std::snprintf(cstr.get(), n + 1, fmt, args...);
   std::string str(cstr.get());
   *toAI << str;
   if (stdinLogStream.get() != nullptr) {
@@ -69,13 +71,13 @@ void flushToAI(std::unique_ptr<boost::process::opstream>& toAI, std::shared_ptr<
   if (stdinLogStream.get() != nullptr) stdinLogStream->flush();
 }
 
-static void logging(std::promise<void> promise, std::unique_ptr<std::istream> input, std::shared_ptr<std::ostream> output, int MAX_SIZE) {
+static void logging(std::promise<void> promise, std::unique_ptr<std::istream> input, std::shared_ptr<std::ostream> output, std::shared_ptr<std::mutex> mutex, int MAX_SIZE) {
   if (output) {
     int size = 0;
     for (; MAX_SIZE == -1 || size < MAX_SIZE; ++size) {
       char c;
       if (input->get(c)) {
-        std::lock_guard<std::mutex> lock(mutex);
+        std::lock_guard<std::mutex> lock(*mutex);
         *output << c;
       } else {
         break;
@@ -89,14 +91,14 @@ static void logging(std::promise<void> promise, std::unique_ptr<std::istream> in
   promise.set_value();
 }
 
-Logger::Logger(std::unique_ptr<std::istream> input, std::shared_ptr<std::ostream> output, int MAX_SIZE = -1) {
+Logger::Logger(std::unique_ptr<std::istream> input, std::shared_ptr<std::ostream> output, int MAX_SIZE = -1) : mutex(new std::mutex) {
   std::promise<void> promise;
   future = promise.get_future();
-  thread = std::thread(logging, std::move(promise), std::move(input), output, MAX_SIZE);
+  thread = std::thread(logging, std::move(promise), std::move(input), output, mutex, MAX_SIZE);
 }
 
 Logger::~Logger() {
-  mutex.unlock();
+  mutex->unlock();
   future.wait_for(std::chrono::milliseconds(100));
   thread.join();
 }
@@ -140,7 +142,7 @@ Player::Player(string command, const RaceCourse &course, int xpos,
   auto timeUsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
   timeLeft -= timeUsed;
   thread.join();
-  stderrLogger->mutex.lock();
+  stderrLogger->mutex->lock();
   if (stderrLogStream) {
     *stderrLogStream << "[system] spend time: " << timeUsed << ", remain: " << timeLeft << endl;
   }
@@ -227,7 +229,7 @@ IntVec Player::play(int c, Player &op, RaceCourse &course) {
     *stderrLogStream << "[system] turn: " << c << std::endl;
   }
   sendToAI(toAI, stdinLogStream, "%d\n", c);
-  sendToAI(toAI, stdinLogStream, "%d\n", timeLeft);
+  sendToAI(toAI, stdinLogStream, "%" PRId64 "\n", timeLeft);
   sendToAI(toAI, stdinLogStream, "%d ", position.x);
   sendToAI(toAI, stdinLogStream, "%d ", position.y);
   sendToAI(toAI, stdinLogStream, "%d ", velocity.x);
@@ -248,7 +250,7 @@ IntVec Player::play(int c, Player &op, RaceCourse &course) {
   flushToAI(toAI, stdinLogStream);
   std::promise<std::pair<std::unique_ptr<boost::process::ipstream>, Message4Act>> promise;
   std::future<std::pair<std::unique_ptr<boost::process::ipstream>, Message4Act>> future = promise.get_future();
-  stderrLogger->mutex.unlock();
+  stderrLogger->mutex->unlock();
   if (resumeCommand) {
     std::error_code ec;
     int result = boost::process::system(resumeCommand.get(), ec);
@@ -262,7 +264,7 @@ IntVec Player::play(int c, Player &op, RaceCourse &course) {
   auto timeUsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
   timeLeft -= timeUsed;
   thread.join();
-  stderrLogger->mutex.lock();
+  stderrLogger->mutex->lock();
   if (stderrLogStream) {
     *stderrLogStream << "[system] spend time: " << timeUsed << ", remain: " << timeLeft << endl;
   }

--- a/official/player.hpp
+++ b/official/player.hpp
@@ -26,7 +26,7 @@ private:
   std::thread thread;
   std::future<void> future;
 public:
-  std::mutex mutex;
+  std::shared_ptr<std::mutex> mutex;
   Logger(std::unique_ptr<std::istream> input, std::shared_ptr<std::ostream> output, int MAX_SIZE);
   ~Logger();
 };
@@ -38,7 +38,7 @@ struct Player {
   string name;
   Point position;
   IntVec velocity;
-  int timeLeft;
+  int64_t timeLeft;
   Status status;
   std::shared_ptr<std::ofstream> stdinLogStream;
   std::shared_ptr<std::ofstream> stderrLogStream;

--- a/player/player_greedy.vcxproj
+++ b/player/player_greedy.vcxproj
@@ -91,14 +91,14 @@
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>Disabled</Optimization>
       <SDLCheck>true</SDLCheck>
     </ClCompile>
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>MaxSpeed</Optimization>
       <FunctionLevelLinking>true</FunctionLevelLinking>
       <IntrinsicFunctions>true</IntrinsicFunctions>

--- a/player/player_pseudo_alphabeta.vcxproj
+++ b/player/player_pseudo_alphabeta.vcxproj
@@ -96,7 +96,7 @@
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>Disabled</Optimization>
       <SDLCheck>true</SDLCheck>
     </ClCompile>
@@ -110,7 +110,7 @@
   </ItemDefinitionGroup>
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
     <ClCompile>
-      <WarningLevel>Level3</WarningLevel>
+      <WarningLevel>EnableAllWarnings</WarningLevel>
       <Optimization>MaxSpeed</Optimization>
       <FunctionLevelLinking>true</FunctionLevelLinking>
       <IntrinsicFunctions>true</IntrinsicFunctions>

